### ISSUE TITLE
Tracking: Upgrade to react-native@0.8

### ISSUE THREAD

**Issue by KevinGrandon** [alinz/react-native-webview-bridge#12]
It would be nice to be able to upgrade to react-native@0.8. I'm opening this to track the issue, and potentially investigate fixes.

There is an issue filed in the react-native project at: https://github.com/facebook/react-native/issues/2010

In the meantime, we may be able to implement a retry workaround. @lightsofapollo commented there with, 
> Okay after a minimum of digging it turns out that it's a fairly simple race... It looks like pre 0.7 attempting to access viewRegistry for a particular reactTag was more deterministic (or the race was smaller) ... In my case I wrap my access to viewRegistry in a retry (which itself is only called in addUIBlock blocks).

**Comment by alinz** [comment]
Thanks @KevinGrandon. At the moment I think the current version can be compiled against 0.8 version of react-native. However, because of that issue, you will be stuck!

I can't wait to opgrade, but because my app is depends on WebView heavily, I can not risk it. I just have to wait a little bit longer. I might end up fixing the react-native bug if I find some time off in the weekend!

**Comment by KevinGrandon** [comment]
@alinz - When trying to upgrade to 0.8 locally, I ran into this error message: `'Error: undefined is not a function (evaluating \'this.getWebWiewHandle()\')\n stack: \n  getWebViewBridgeHandler`

This seems a bit different than the one reported in the react-native bug. Were you seeing this error message as well?

**Comment by alinz** [comment]
@KevinGrandon My current app is heavily dependent on WebView. I have been blocked by that issue 2010. I really want to continue but until this bug gets fix, there is no point to upgrade to 0.8. 

However, I will give it another shot by end of today or tomorrow. I will let you know.

btw, thank you for helping me out.

**Comment by skitterm** [comment]
There is a typo in the jsx script file -- it says "getWebWiewHandle" instead of "getWebViewHandle". Didn't solve my problem, but nonetheless...

**Comment by alinz** [comment]
@skitterm this is really funny! until 0.8-rc2, it was `getWebWiewHandle` and they have change it to `getWebViewHandle` at [this commit](https://github.com/facebook/react-native/commit/bb141e3a3c14d45e56fd897499b8aad1a1e7d814#diff-eb29dcb037331f7852621383e484d706L182). Thanks for pointing this out.

As far as I know, the WebView bug still there. I'm trying to find some time and find the actual bug in react-native. Still not find the time. I will let you know soon!

**Comment by skitterm** [comment]
Sounds good. A little more digging on my part: I found that on ComponentDidMount, WebViewManager didn't have the bridgeSetup method attached (the only methods that were there were goBack, goForward, and reload, and the NavigationType property).

**Comment by alinz** [comment]
OK, I have released a new version which fixes the crash issue in react-native 0.7 above. Please upgrade to 0.3.2 and let me know if you have this issue.

cc: @skitterm @KevinGrandon

**Comment by alinz** [comment]
I'm going to close this issue since I have tested with 0.8 and above. One thing you guys have to be aware is that, 0.9 has `-rc` in there version number which causes npm to throw an error. For time being, you can fork the react-antive and edit the package.json or fork this project and added -rc to dependencies value.

**Comment by skitterm** [comment]
It looks like I still have the same error: undefined is not a function (evaluating WebViewHandler.bridgeSetup(handlerId)) at WebViewBridge, line 76. 

I'm using React-Native 0.8.0 and version 0.3.2 of this library (and the newest version of the react-native-cli). I created a new project with the CLI, linked up the library, added in a basic WebViewBridge, and ran it. The above error was what I got. I had the same issue with the example in the node_modules folder.

**Comment by alinz** [comment]
@skitterm this only happens if you don't include the objective-c into your project. 
I have tested the Sample project and it seems working. So I think you missed one step.

Here's my approach to install this library.

1: add the `react-native-webview-bridge` in your project's package.json.
2: npm install
3: go to xcode and right click on Libraries
4: select `Add files to "..."`
5: go to node_modules/react-native-webview-bridge and select `WebViewBridge`
make sure that you see your Libraries folder like this:

![screen shot 2015-08-04 at 2 48 27 pm](https://cloud.githubusercontent.com/assets/967050/9069743/05377424-3ab9-11e5-84d9-666035dcfd10.png)

 compile and run the app.

**Comment by alinz** [comment]
One more thing, I have updated the version to 0.3.3, I found the bug and I fixed it.

**Comment by seanchenca** [comment]
It seems that the "undefined is not a function (evaluating WebViewManager.bridgeSetup(this.getWebViewBridgeHandler());" is coming back. I am using 0.3.5 and React Native 0.9.
<img width="376" alt="screen shot 2015-08-21 at 9 35 37 pm" src="https://cloud.githubusercontent.com/assets/561815/9422550/9e299dba-484c-11e5-8f27-53052b116150.png">

**Comment by alinz** [comment]
I've been using this in my app with react-native 0.10 and I have not seen this issue. Can you provide me some sample code? then I can help you with this.

**Comment by seanchenca** [comment]
Hi alinz, here is my index.io.js:
```
/**
 * Sample React Native App
 * https://github.com/facebook/react-native
 */
'use strict';

var React = require('react-native');
var WebViewBridgeNative = require('react-native-webview-bridge');

var myPageScripts = function() {
  function WebViewBridgeReady(cb) {
    if (window.WebViewBridge) {
      cb(window.WebViewBridge);
      return;
    }
    function handler() {
      document.removeEventListener('WebViewBridge', handler, false);
      cb(window.WebViewBridge);
    }
    document.addEventListener('WebViewBridge', handler, false);
  }
  WebViewBridgeReady(function (WebViewBridge) {
    WebViewBridge.send("Hello this is me calling from web page");
  });
};

//convert function definition into string
myPageScripts = `(${myPageScripts.toString()}());`;

var {
  AppRegistry,
  Component,
  StyleSheet,
  Text,
  View,
} = React;

class WVTest extends Component {
  constructor(props) {
    super(props);
    this.once = true;
  }
  
  componentDidMount() {
    var myWebViewBridgeRef = this.refs.myWebViewBridge;

    var onMessage = function(message) {
      console.log("Received message", message);
    };
    myWebViewBridgeRef.onMessage(onMessage);

    myWebViewBridgeRef.injectBridgeScript();

    //this opens up the printer window dialoge
    setTimeout(() => {
      myWebViewBridgeRef.print();
    }, 5000);
  }

  onNavigationStateChange(navState) {
    var myWebViewBridgeRef = this.refs.myWebViewBridge;
    if (this.once) {
      this.once = false;
      setTimeout(() => {
        myWebViewBridgeRef.evalScript(myPageScripts);
      }, 1000);
    }

    console.log(navState.url);
  }

  render() {
    var url = 'http://google.com';

    return (
      <WebViewBridgeNative
        ref="myWebViewBridge"
        onNavigationStateChange={this.onNavigationStateChange.bind(this)}
        url='web/index.html'
        style={{flex: 1}}/>
    );
  }
}

var styles = StyleSheet.create({
  container: {
    flex: 1,
    justifyContent: 'center',
    alignItems: 'center',
    backgroundColor: '#F5FCFF',
  },
  welcome: {
    fontSize: 20,
    textAlign: 'center',
    margin: 10,
  },
  instructions: {
    textAlign: 'center',
    color: '#333333',
    marginBottom: 5,
  },
});

AppRegistry.registerComponent('WVTest', () => WVTest);
```
I have try leaving url: {url} in the render() function but it's the same. Here is the index.html under web/ :
```
<!DOCTYPE html>
<html>
  <head>
    <meta charset="UTF-8" />
  </head>
  <body style="margin: 0; padding: 0; height: 300px">
    <h1>Hi</h1>
    <canvas style="width:100%"></canvas>
  </body>
  <script type="text/javascript">
  function WebViewBridgeReady(cb) {

  	if (window.WebViewBridge) {
      cb(window.WebViewBridge);
      return;
  	}

  	function handler() {
      //remove the handler from listener since we don't need it anymore
      document.removeEventListener('WebViewBridge', handler, false);
      //pass the WebViewBridge object to the callback
      cb(window.WebViewBridge);
    }

    //if WebViewBridge doesn't exist in global scope attach itself to document
    //event system. Once the code is being injected by extension, the handler will
    //be called.
    document.addEventListener('WebViewBridge', handler, false);
  };

  //WebViewBridgeReady(function(webViewBridge) {
  //	webViewBridge.onMessage = function(message) {
  //	  console.log(message);
  //  };
  //});
  </script>
</html>
```
After I click on Dismiss (ESC) the page does show up (displaying "Hi"). Below is the screen shot of my project. Thanks for looking into it.

<img width="257" alt="screen shot 2015-08-22 at 12 38 47 pm" src="https://cloud.githubusercontent.com/assets/561815/9425608/d4dd3564-48ca-11e5-9c77-69bc59c49757.png">

**Comment by KevinGrandon** [comment]
You might need to add the files individually to the Libraries folder (instead of adding the entire folder).

**Comment by seanchenca** [comment]
Have just tried that. Now got these compile error. Not sure why Xcode doesn't recognize nonnull?

<img width="1021" alt="screen shot 2015-08-22 at 5 13 12 pm" src="https://cloud.githubusercontent.com/assets/561815/9426364/28255820-48f1-11e5-812f-edd49f3d1674.png">

**Comment by alinz** [comment]
Remove the nonnull if you are using react-native under 0.9.

**Comment by seanchenca** [comment]
Wow, that does the trick! The error is gone and it works now. Thanks for the help!

**Comment by alinz** [comment]
`nonnull` was introduced to the react-native api to compatible with Android version. Always read the release notes from the api to make sure you don't break any changes. here's the [link](https://github.com/facebook/react-native/releases)

**Comment by seanchenca** [comment]
I think I am using react-native 0.9.0. Should I add the nonnull only after upgrade to 0.10.0?

**Comment by alinz** [comment]
@seanchenca Yes! based on release note, you should use `nonnull` after upgrading to 0.10.x

**Comment by seanchenca** [comment]
Thanks for the info alinz. Will remember to add them back once I upgrade.

**Comment by johanmontan** [comment]
Had the same issues even thought I followed the instructions in  [Linking Libraries(]http://facebook.github.io/react-native/docs/linking-libraries-ios.html) as it was my first react-native project and didn't have a clue where to begin.
I realised that the this library didn't build properly so I commented out the code that didn't work. In this case RCTLogMustFix(). 
(Running in react-native 0.15.0 within Xcode 7.1).
